File: archive/cold-storage/reference/aaes_os_v1/typescript/src/trace_bus.ts
import type { GovernedSpan, TraceBus, TraceEvent } from "./types.js";
import { err, notImplemented } from "./error.js";

export class TraceBusStub implements TraceBus {
  private readonly log: TraceEvent[] = [];
  private readonly spans = new Map<string, GovernedSpan>();

  validate(event: TraceEvent, span: GovernedSpan) {
    if (event.span_id !== span.span_id) {
      return err<TraceEvent>("AAES_SPAN_STATE_INVALID", "event.span_id does not match span");
    }
    return notImplemented<TraceEvent>("TraceBus.validate");
  }

  append(event: TraceEvent, span: GovernedSpan) {
    const validated = this.validate(event, span);
    if (!validated.ok) {
      return validated;
    }
    this.log.push(event);
    this.spans.set(span.span_id, span);
    return validated;
  }

  get_events(span_id: string): TraceEvent[] {
    return this.log.filter((event) => event.span_id === span_id);
  }

  validate_and_append(event: TraceEvent, span: GovernedSpan) {
    return this.append(event, span);
  }

  register_span(span: GovernedSpan): void {
    this.spans.set(span.span_id, span);
  }
}

File: archive/cold-storage/reference/aaes_os_v1/rust/src/trace_bus.rs
use crate::error::{not_implemented, AaesResult};
use crate::governed_span::GovernedSpan;
use crate::types::TraceEvent;

#[derive(Debug, Default)]
pub struct TraceBusStub {
    log: Vec<TraceEvent>,
    spans: std::collections::HashMap<String, GovernedSpan>,
}

impl TraceBusStub {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn validate(&self, event: &TraceEvent, span: &GovernedSpan) -> AaesResult<TraceEvent> {
        if event.span_id != span.span_id {
            return Err(crate::error::aaes_err(
                "AAES_SPAN_STATE_INVALID",
                "event.span_id does not match span",
            ));
        }
        Err(not_implemented("TraceBus.validate"))
    }

    pub fn append(&mut self, event: TraceEvent, span: &GovernedSpan) -> AaesResult<TraceEvent> {
        let _ = self.validate(&event, span)?;
        self.log.push(event.clone());
        self.spans.insert(span.span_id.clone(), span.clone());
        Ok(event)
    }

    pub fn get_events(&self, span_id: &str) -> Vec<TraceEvent> {
        self.log
            .iter()
            .filter(|event| event.span_id == span_id)
            .cloned()
            .collect()
    }

    pub fn validate_and_append(
        &mut self,
        event: TraceEvent,
        span: &GovernedSpan,
    ) -> AaesResult<TraceEvent> {
        self.append(event, span)
    }

    pub fn register_span(&mut self, span: GovernedSpan) {
        self.spans.insert(span.span_id.clone(), span);
    }
}
